// next/dist/build/webpack/loaders/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.buildExports = buildExports;
exports.loadEdgeFunctionConfigFromFile = loadEdgeFunctionConfigFromFile;
exports.serverComponentRegex = exports.clientComponentRegex = exports.defaultJsFileExtensions = void 0;
var _getPageStaticInfo = require("../../analysis/get-page-static-info");
const defaultJsFileExtensions = [
    "js",
    "mjs",
    "jsx",
    "ts",
    "tsx"
];
exports.defaultJsFileExtensions = defaultJsFileExtensions;
const imageExtensions = [
    "jpg",
    "jpeg",
    "png",
    "webp",
    "avif"
];
const nextClientComponents = [
    "dist/client/link",
    "dist/client/image",
    "dist/client/future/image",
    "dist/shared/lib/head",
    "dist/client/script",
    "dist/shared/lib/dynamic", 
];
function buildExports(moduleExports, isESM) {
    let ret = "";
    Object.keys(moduleExports).forEach((key)=>{
        const exportExpression = isESM ? `export ${key === "default" ? key : `const ${key} =`} ${moduleExports[key]}` : `exports.${key} = ${moduleExports[key]}`;
        ret += exportExpression + "\n";
    });
    return ret;
}
const clientComponentRegex = new RegExp("(" + `\\.client(\\.(${defaultJsFileExtensions.join("|")}))?|` + `next[\\\\/](${nextClientComponents.join("|")})(\\.js)?|` + `\\.(${imageExtensions.join("|")})` + ")$");
exports.clientComponentRegex = clientComponentRegex;
const serverComponentRegex = new RegExp(`\\.server(\\.(${defaultJsFileExtensions.join("|")}))?$`);
exports.serverComponentRegex = serverComponentRegex;
async function loadEdgeFunctionConfigFromFile(absolutePagePath, resolve) {
    const pageFilePath = await resolve("/", absolutePagePath);
    return (await (0, _getPageStaticInfo).getPageStaticInfo({
        nextConfig: {},
        pageFilePath,
        isDev: false
    })).middleware;
}
