// nextjs-progressbar/dist/index.js
var N=Object.create;var a=Object.defineProperty;var k=Object.getOwnPropertyDescriptor;var v=Object.getOwnPropertyNames;var E=Object.getPrototypeOf,$=Object.prototype.hasOwnProperty;var y=e=>a(e,"__esModule",{value:!0}),g=(e,r)=>a(e,"name",{value:r,configurable:!0});var S=(e,r)=>{for(var o in r)a(e,o,{get:r[o],enumerable:!0})},T=(e,r,o,l)=>{if(r&&typeof r=="object"||typeof r=="function")for(let s of v(r))!$.call(e,s)&&(o||s!=="default")&&a(e,s,{get:()=>r[s],enumerable:!(l=k(r,s))||l.enumerable});return e},m=(e,r)=>T(y(a(e!=null?N(E(e)):{},"default",!r&&e&&e.__esModule?{get:()=>e.default,enumerable:!0}:{value:e,enumerable:!0})),e),_=(e=>(r,o)=>e&&e.get(r)||(o=T(y({}),r,1),e&&e.set(r,o),o))(typeof WeakMap!="undefined"?new WeakMap:0);var z={};S(z,{default:()=>b});var i=m(require("next/router")),n=m(require("nprogress")),t=m(require("prop-types")),f=m(require("react"));function b({color:e="#29D",startPosition:r=.3,stopDelayMs:o=200,height:l=3,showOnShallow:s=!0,options:d,nonce:w}){let p=null;f.useEffect(()=>(d&&n.configure(d),i.default.events.on("routeChangeStart",h),i.default.events.on("routeChangeComplete",x),i.default.events.on("routeChangeError",c),()=>{i.default.events.off("routeChangeStart",h),i.default.events.off("routeChangeComplete",x),i.default.events.off("routeChangeError",c)}),[]);let h=g((P,{shallow:u})=>{(!u||s)&&(n.set(r),n.start())},"routeChangeStart"),x=g((P,{shallow:u})=>{(!u||s)&&(p&&clearTimeout(p),p=setTimeout(()=>{n.done(!0)},o))},"routeChangeEnd"),c=g((P,u,{shallow:C})=>{(!C||s)&&(p&&clearTimeout(p),p=setTimeout(()=>{n.done(!0)},o))},"routeChangeError");return f.createElement("style",{nonce:w},`
      #nprogress {
        pointer-events: none;
      }
      #nprogress .bar {
        background: ${e};
        position: fixed;
        z-index: 9999;
        top: 0;
        left: 0;
        width: 100%;
        height: ${l}px;
      }
      #nprogress .peg {
        display: block;
        position: absolute;
        right: 0px;
        width: 100px;
        height: 100%;
        box-shadow: 0 0 10px ${e}, 0 0 5px ${e};
        opacity: 1;
        -webkit-transform: rotate(3deg) translate(0px, -4px);
        -ms-transform: rotate(3deg) translate(0px, -4px);
        transform: rotate(3deg) translate(0px, -4px);
      }
      #nprogress .spinner {
        display: block;
        position: fixed;
        z-index: 1031;
        top: 15px;
        right: 15px;
      }
      #nprogress .spinner-icon {
        width: 18px;
        height: 18px;
        box-sizing: border-box;
        border: solid 2px transparent;
        border-top-color: ${e};
        border-left-color: ${e};
        border-radius: 50%;
        -webkit-animation: nprogresss-spinner 400ms linear infinite;
        animation: nprogress-spinner 400ms linear infinite;
      }
      .nprogress-custom-parent {
        overflow: hidden;
        position: relative;
      }
      .nprogress-custom-parent #nprogress .spinner,
      .nprogress-custom-parent #nprogress .bar {
        position: absolute;
      }
      @-webkit-keyframes nprogress-spinner {
        0% {
          -webkit-transform: rotate(0deg);
        }
        100% {
          -webkit-transform: rotate(360deg);
        }
      }
      @keyframes nprogress-spinner {
        0% {
          transform: rotate(0deg);
        }
        100% {
          transform: rotate(360deg);
        }
      }
    `)}g(b,"NextNProgress");b.propTypes={color:t.string,startPosition:t.number,stopDelayMs:t.number,height:t.number,showOnShallow:t.bool,options:t.object,nonce:t.string};module.exports=_(z);0&&(module.exports={});
